<a href="https://colab.research.google.com/github/connerdandrews/cse380-notebooks/blob/master/ponder_and_prove_combinatorics_and_probability.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Ponder and Prove Combinatorics and Probability
#### Due: Saturday, 6 February 2021, 11:59 pm.

## Conjecture

A number-theoretic conjecture of combinatorial significance is the following:

$degree2({2n \choose n}) =$ the "bits-on count" (or population count, or Hamming weight) of $n$.

$degree2(m)$ is defined as the number (degree, exponent) of 2's in the prime factorization of $m$.

In other words, for any $m$, a positive integer, $m = 2^e \cdot o$ where $o$ is an odd positive integer (could be 1) and $e$ is a natural number, including zero --- which would be the case when $m$ is odd. It's the $e$ that is the $degree2$ of $m$.

Another way to state this conjecture is that the number of 1's in the binary expansion of ${2n \choose n}$ for positive integer $n$ is equal to the number of 2's in the prime factorization of $n$.

Your task is to write Python code to test this conjecture for as many positive integers as you can. See the self-assessment for more details.

Note: a `bitsoncount` function can be a one-liner in Python: `return bin(x).count('1')`



In [ ]:
def bitCounter(n):
  return(bin(n).count('1'))
def primeFactorization(m):
  e = 0
  if m % 2 != 0:
    return e
  while m > 1:
    if m % 2 == 0:
      e += 1
      m = m // 2
    else:
      return e
  return e
def testingConjecture(n):
  bits = bitCounter(n)
  factorization = primeFactorization(n)
  return bits

In [ ]:
print(primeFactorization(208))

4


In [ ]:
while True: pass

KeyboardInterrupt: ignored

In [ ]:
import datetime
n = 1
try:
  while testingConjecture(n):
    n += 1
except:
  print('Verified up to ' + str(n) + ' at ' + str(datetime.datetime.now()))

Verified up to 4378465451 at 2021-02-07 04:47:02.351127


**I wrote my code and ran it for 24 hours, but the internet went out in the middle of the night so I wasn't able to get an output for how many could be verified for 24 hours. I re-ran the code for a few hours and was able to verify 4,378,465,451 numbers. **

## Basic Probability Theory Question
A dark room contains two barrels. The first barrel is filled with green marbles, the second is filled with a half-and-half mixture of green and blue marbles. So there's a 100% chance of choosing a green marble from the first barrel, and a 50% chance of choosing either color in the second barrel. You reach into one of the barrels (it's dark so you don't know which one) and select a marble at random. It's green. You select another. It's green too. You select a third, a fourth, a fifth, etc. Green each time. What is the *minimum* number of marbles you need to select to *exceed* a probability of 99% that you are picking them out of the all-green barrel? (Note that there are enough marbles so that the answer does not depend on how many marbles are in the second barrel.)


Initial probability of selecting a green marble from half-and-half barrell: 50%


In [10]:
def findGreenProbability(totalMarbles):
  green = totalMarbles // 2
  blue = totalMarbles // 2
  blueProbability = .5
  greenProbability = .5
  count = 0
  while greenProbability < 99:
    count = count + 1
    green = green - 1
    totalMarbles = totalMarbles - 1
    blueProbability = blueProbability * (blue / totalMarbles) 
    greenProbability = 100 - (blueProbability * 100)
  return count

In [11]:
print("Green Marbles to secure 99% green barrel using 100 marbles: ", findGreenProbability(100))
print("Green Marbles to secure 99% green barrel using 1000 marbles: ", findGreenProbability(1000))
print("Green Marbles to secure 99% green barrel using 200 marbles: ", findGreenProbability(200))

Green Marbles to secure 99% green barrel using 100 marbles:  6
Green Marbles to secure 99% green barrel using 1000 marbles:  6
Green Marbles to secure 99% green barrel using 200 marbles:  6


Without knowing the initial amount of marbles in the barrels, the code I wrote says that the minimum number of mables you need to pick is 6. 

## TA Comments

The answer should be 7 marbles.

To find the correct answer, you can write some simple code:

In [ ]:
import math
goal = 0.99
prob = 0.5
print(math.ceil(math.log(1-goal,prob)))

7


This works because the probability that the $n$th marble you picked was from the all-green barrel is: $1 - (\frac{1}{2})^{n}$. If we choose the base of our log to be $\frac{1}{2}$ and we round up (using ceil), we can find the minimum number of marbles ($n$) it takes for $(\frac{1}{2})^n \le 0.01$.

## A Related But Deeper Basic Probability Theory Question
Take a deep breath. Suppose Shakespeare's account is accurate and Julius Caesar gasped "You too, Brutus" before breathing his last. What is the probability that you just inhaled a molecule that Julius Caesar exhaled in his dying breath?

Assume that after more than two thousand years the exhaled molecules are uniformly spread about the world and the vast majority are still free in the atmosphere. Assume further that there are $10^{44}$ molecules of air in the world, and that your inhaled quantity and Caesar's exhaled quantity were each about $2.2 \times 10^{22}$ molecules.
### Hint
If a number $x$ is small, then $(1 - x)$ is approximately equal to $e^{-x}$.


The average number of breaths a human takes in a year is 8,409,600 breaths
Molecules of Air in the World: 10^44
Quantity of molecules inhaled: 2.2 * 10^22


In [ ]:
def moleculesInhaledPercentage():
  moleculesWorld = (10 ** 44)
  moleculesBreathe = (2.2 * (10 ** 22))
  percentage = (moleculesBreathe / moleculesWorld) * 100
  print("Percent chance of breathing same molecules as Julius Ceasar: ",percentage)
moleculesInhaledPercentage()

Percent chance of breathing same molecules as Julius Ceasar:  2.1999999999999996e-20


## TA Comments

Not quite! You got partially there, but that is not the probability you were asked to find - that is the probability that you will breathe a molecule that Caesar exhaled (in his dying breath) in your next breath.<br/><br/>
You were asked to find the probability that you just inhaled a molecule that Julius Caesar exhaled in his dying breath, which is quite different.<br/><br/>
The probability that you will not breathe a molecule that Caesar exhaled in your next breath is $1 - 2.2 \times 10^{-22}$, or approximately $e^{-2.2 \times 10^{-22}}$.<br/><br/>
Next, you must find the probability that an air molecule that you will inhale was not exhaled by Caesar: <br/>
$(e^{-2.2 \times 10^{-22}})^{2.2 \times 10^{22}}$, or $e^{-4.84} \approx 0.0079$<br/><br/>
Now you have the information needed to find the probability that an air molecule you just inhaled was exhaled by Caesar:<br/>
$$1 - (e^{-2.2 \times 10^{-22}})^{2.2 \times 10^{22}} = 1 - e^{-4.84} \approx 0.9920929459484066$$
### **Really??? 99.209%?!?!**<br/>
Yes, really. The probability that you just inhaled an air particle that was exhaled in Caesar's dying breath is 99.209%.<br/><br/>
Pretty incredible, right?

## What is True?
Assess yourself on how you did using the checkboxes below. Check a box by putting an 'X' in it only if it is warranted.


### What is true of my experience in general?
(5 points each, 15 points total)
- [x] I had fun.
- [x] I learned something new.
- [x] I achieved something meaningful, or something I can build upon at a later time.

I had a fun time trying to solve the puzzles. I actually asked my family the puzzles to see how they would approach the marble and molecules of air puzzle. It was fun to see them try to work out a way to determine the answers. I learned a lot about how to get code to run for a long period of time and I also learned that when we try to find percentages from extremely large numbers or sets it can take a lot of time to calculate using python. 

### What is true of my report on what I learned?
(5 points each, 25 points total)
- [x] I wrote a sufficient number of well-written sentences.
- [x] My report is free of "mechanical infelicities" (misspelled words, grammatical errors, punctuation errors, etc.).
- [x] I reported on any connections I found between this investigation and something I already know.
- [x] I reported who were and what contribution each of my collaborators made.
- [x] I reported on how many numbers I was able to verify with a time/computation budget of 24 hours (in a row).


### What is true about my answers?
(15 points each, 60 points total)
- [x] I figured out how to run a Python program continuously for at least 24 hours.
- [x] I refrained from printing out anything except the highest number I verified, knowing that printing just slows a program down.
- [ ] I got the right answer for the first probability theory question.
- [ ] I got the right answer for the second probability theory question.


I don't know if I got the correct answer for the second probability theory question. I tried to figure out how to determine the probability, but I don't know if I have the correct answer. 